### ISSUE TITLE
Support for upsert operations via sqlalchemy-hana?

### ISSUE THREAD

**Issue by therhaag** [SAP/sqlalchemy-hana#164]
Hi - knowing that HANA supports an UPSERT statement, I was wondering whether there is a matching construct in the dialect that would enable this operation via sql_alchemy as it is the case for specific dialects included with sql_alchemy (https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-upsert-statements).

I have used the corresponding functionality for postgresql (https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert) in the past and would now like to port some of this code.

Could you please point me to any resource on how to accomplish that or do I have to fall back to constructing the query manually?
Any help is appreciated - Jan

**Comment by kasium** [comment]
Hi, I guess beside not using UPSERT but a kind of UPDATE/SELECT/INSERT combination with exception handling, I'm not aware of other options than adding support for UPSERT in sqlalchemy-hana. I'm currently busy working on cleanup tasks, but I can have a look in the following weeks

**Comment by kasium** [comment]
@therhaag I did some research and I guess we need to make a difference between SQLAlchemy core and SQLAlchemy ORM.
For SQLAlchemy core, the only way is a manual approach using SELECT, INSERT and UPDATE. Not nice and it has some roundtrips. For this, we could build an UPSERT construct.

For SQLAlchemy ORM using the Session API on the other hand it's not quite easy, because how sessions work. For this, I would recommend using `session.merge`.  I also found [a blog](https://olegkhomenko.medium.com/how-to-create-an-sql-table-and-perform-upsert-using-sqlalchemy-orm-in-3-minutes-3f346ef7bdb5) explaining the idea.

So let's focus on SQLAlchemy Core. I guess it makes sense to invent a completly new construct, because neither insert nor update have a similar syntax compared to UPSERT.

As said, I'm happy to work on this in the next weeks but to be transparent, It will most likly somewhen in January
